(GH-1957) Projects and project directories (#1976)
puppetlabs/bolt: documentation/bolt_project_directories.md

  a `bolt-project.yaml` file to it and configuring the `modulepath` to match the
  `modulepath` in `environment.conf`.

## Embedded project directories

Bolt treats a directory containing a subdirectory named `Boltdir` as an embedded
project directory. Use this type of directory to embed Bolt management code into
another repo.

For example, you can store Bolt management code in the same repo as the
application that Bolt manages, without cluttering up the top level with multiple
files. This structure allows you to run Bolt from anywhere in the application's
directory structure.

An embedded Bolt directory looks like this:

```console
project/
├── Boltdir
│   ├── bolt-project.yaml
│   ├── inventory.yaml
│   ├── plans
│   │   ├── deploy.pp
│   │   └── diagnose.pp
│   └── tasks
│       ├── init.json
│       └── init.py
├── src #non Bolt source code for the project
└── tests #non Bolt tests for the project
```

As long as your `bolt-project.yaml` file contains a `name` field, Bolt loads
your local Bolt content from the top level of the `Boltdir`. Your `Boltdir` can
also contain modules. For more information, see [Modules](modules.md).

> ## How Bolt chooses a Bolt directory

Reviewer: ```suggestion
## How Bolt chooses a project directory
```

├── src #non Bolt source code for the project
└── tests #non Bolt tests for the project
```

> **Note:** If a directory contains both `Boltdir` and `bolt.yaml` or
  `bolt-project.yaml`, the `Boltdir` directory is used as the project directory
  rather then the parent.

### User project directory
As long as your `bolt-project.yaml` file contains a `name` field, Bolt loads
your local Bolt content from the top level of the `Boltdir`. Your `Boltdir` can
also contain modules. For more information, see [Modules](modules.md).

If Bolt can't find a project directory based on `Boltdir` or `bolt.yaml`, it
uses `~/.puppetlabs/bolt` as the project directory. Use this type of directory
if you have a single set of Bolt code and data that you use across all projects.
## How Bolt chooses a Bolt directory

## How the project directory is chosen
If Bolt can't find a directory based on `Boltdir` or `bolt-project.yaml`, it
uses the default: `~/.puppetlabs/bolt-project.yaml`.

Bolt uses these methods, in order, to choose a project directory.
Bolt uses the following methods, in order, to choose a Bolt directory.

1. **Environment variable:** You can specify a path to a project using the

>    `BOLT_PROJECT` environment variable.

Reviewer 2: ```
1. **Environment variable:** You can specify a path to a project using the
   `BOLT_PROJECT` environment variable.
```
This should be added back in.

Author: Ah good catch - I think I missed this when I was resolving a conflict

>    a parent directory, Bolt uses `~/.puppetlabs/bolt-project.yaml`.

Reviewer 2: ```suggestion
   a parent directory, Bolt uses `~/.puppetlabs/bolt/` as the project directory.
```

# Project directories

There are two directory structures available to Bolt projects. You can run Bolt inside a
local project directory and share that directory with others using a version control
system like git, or you can embed the project directory inside an application's
repo, and make Bolt available to that application.

## Local project directories

To track and share management code in a dedicated repository, or to run and
develop Bolt content locally, use a local project directory. Bolt recognizes
any directory containing a `bolt-project.yaml` file as a Bolt project directory.

A simple project directory looks like this:

```console
myproject/
├── bolt-project.yaml
├── inventory.yaml
├── plans
│   ├── deploy.pp
│   └── diagnose.pp
└── tasks
    ├── init.json
    └── init.py
```

As long as your `bolt-project.yaml` file contains a `name` field, Bolt loads
your local Bolt content from the top level of your directory. If you're
developing a module for the Puppet Forge, you can use a Puppet module
directory structure. For more information, see [Module
structure](module_structure.md). 

> > 🔩 **Tip:** You can use an existing control repo as a Bolt directory by adding

Reviewer 2: ```suggestion
> 🔩 **Tip:** You can use an existing Puppet control repo as a Bolt directory by adding
```

I think clarifying that this is for existing Puppet users might be helpful, since it can be a bit confusing what we're talking about if you've never used Puppet.

  `modulepath` in `environment.conf`.

## Embedded project directories

Bolt treats a directory containing a subdirectory named `Boltdir` as an embedded
project directory. Use this type of directory to embed Bolt management code into
another repo.

For example, you can store Bolt management code in the same repo as the
application that Bolt manages, without cluttering up the top level with multiple
files. This structure allows you to run Bolt from anywhere in the application's
directory structure.

An embedded Bolt directory looks like this:

```console
project/
├── Boltdir
│   ├── bolt-project.yaml
│   ├── inventory.yaml
│   ├── plans
│   │   ├── deploy.pp
│   │   └── diagnose.pp
│   └── tasks
│       ├── init.json
│       └── init.py
├── src #non Bolt source code for the project
└── tests #non Bolt tests for the project
```

As long as your `bolt-project.yaml` file contains a `name` field, Bolt loads
your local Bolt content from the top level of the `Boltdir`. Your `Boltdir` can
also contain modules. For more information, see [Modules](modules.md).

## How Bolt chooses a project directory

If Bolt can't find a directory based on `Boltdir` or `bolt-project.yaml`, it

> uses the default: `~/.puppetlabs/bolt-project.yaml`.

Reviewer 2: ```suggestion
uses the default: `~/.puppetlabs/bolt/`.
```

```

> **Note:** The `BOLT_PROJECT` environment variable takes precedence over the
> `--configfile` CLI option. 

## Common files and directories

The default paths for all Bolt configuration, code, and data are relative to the
module path.

|Directory|Description|
|---------|-----------|
|[`bolt.yaml`](bolt_configuration_reference.md)|Contains configuration options for Bolt. ⛔ **`bolt.yaml` is deprecated; use `bolt-project.yaml` instead.** |
|`hiera.yaml`|Contains the Hiera config to use for target-specific data when using `apply`.|
|[`inventory.yaml`](inventory_file_v2.md)|Contains a list of known targets and target specific data.|
|[`bolt-project.yaml`](bolt_configuration_reference.md#project_configuration_options)|Contains configuration options for Bolt and Bolt projects. For more information on Bolt projects, see [Bolt projects](./experimental_features.md#bolt-projects).|
|[`Puppetfile`](bolt_installing_modules.md#)|Specifies which modules to install for the project.|
|[`modules/`](bolt_installing_modules.md#)|The directory where modules from the `Puppetfile` are installed. In most cases, do not edit these modules locally.|

> |[`site-modules`](bolt_installing_modules.md)|Local modules that are edited and versioned with the Bolt directory.|

Reviewer 2: ```suggestion
|[`site-modules/`](bolt_installing_modules.md)|Local modules that are edited and versioned with the Bolt directory.|
```
